Lib/_pyio.py

```diff
@@ -2529,3 +2530,23 @@ def encoding(self):
     def detach(self):
         # This doesn't make sense on StringIO.
         self._unsupported("detach")
+
+
+# ____________________________________________________________
+
+import atexit, weakref
+
+_all_writers = weakref.WeakSet()
+
+def _register_writer(w):
+    # keep weak-ref to buffered writer
+    _all_writers.add(w)
+
+def _flush_all_writers():
+    # Ensure all buffered writers are flushed before proceeding with
+    # normal shutdown.  Otherwise, if the underlying file objects get
+    # finalized before the buffered writer wrapping it then any buffered
+    # data will be lost.
+    for w in _all_writers:
+        w.flush()
```

Comment:
Common case for a _flush_ error is a broken pipe or disk full. Try writing to _/dev/full_. It would be nice if this indicated an error, but I suspect it would be hard to do (related: <https://bugs.python.org/issue5319>).